onnxscript/ir/_external_data.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Pass to save tensor data as external tensors."""
+
+import os
+
+from onnxscript.ir import _core
+
+
+class ExternalDataInfo:
+    def __init__(self, name, offset, length):
+        self.name = name
+        self.offset = offset
+        self.length = length
+
+
+def save_external_data(
+    initializers,
+    file_path,
+    allocation_granularity: int = 65536,  # 64KB
+):
+    # Store external data such as name, offset and size
+    external_data_info = dict()
+
+    # Create file if it doesn't exist
+    if not os.path.isfile(file_path):
+        with open(file_path, "ab"):
+            pass
+
+    with open(file_path, "r+b") as data_file:
+        current_offset = 0
+        for i_name, i_value in initializers.items():
+            tensor_val = i_value
+            if isinstance(tensor_val, _core.Value):
+                tensor_val = i_value.const_value()  # pylint: disable=protected-access
+            assert isinstance(tensor_val, _core.Tensor)
+            raw_data = tensor_val.tobytes()
+            tensor_size = tensor_val.size
+            # Convert each initializer to core.ExternalTensor
+            # Align tensors
+            alignment_factor = max(4096, allocation_granularity)
+            current_offset = (
+                (current_offset + alignment_factor - 1) // alignment_factor * alignment_factor
+            )
+
+            data_file.seek(0, 2)
+            if current_offset is not None:
+                # Pad file to required offset if needed
+                file_size = data_file.tell()
+                if current_offset > file_size:
+                    data_file.write(b"\0" * (current_offset - file_size))
+                data_file.seek(current_offset)
+            data_file.write(raw_data)
+
+            # Store tensor external data
+            external_data_info[i_value.name] = ExternalDataInfo(
+                i_name,
+                current_offset,
+                tensor_size,
+            )
+
+            # Update offset
+            current_offset += tensor_size
+    return external_data_info
+
+
+def convert_model_to_external_data(model: _core.Model, base_path: str, file_path: str = ""):
```

Comment:
Always support os.PathLike in addition to str when an argument is a file path. `os.path.join` will handle both

```suggestion
def convert_model_to_external_data(model: _core.Model, base_path: str | os.PathLike, file_path: str = ""):
```